File: src/bmsx/machine/devices/audio/playback.ts
import type { BiquadFilterType } from './biquad_filter';
import { toSignedWord } from '../../common/numeric';
import {
	APU_FILTER_ALLPASS,
	APU_FILTER_BANDPASS,
	APU_FILTER_HIGHPASS,
	APU_FILTER_HIGHSHELF,
	APU_FILTER_LOWSHELF,
	APU_FILTER_NONE,
	APU_FILTER_NOTCH,
	APU_FILTER_PEAKING,
	APU_GAIN_Q12_ONE,
	APU_PARAMETER_FILTER_FREQ_HZ_INDEX,
	APU_PARAMETER_FILTER_GAIN_MILLIDB_INDEX,
	APU_PARAMETER_FILTER_KIND_INDEX,
	APU_PARAMETER_FILTER_Q_MILLI_INDEX,
	APU_PARAMETER_GAIN_Q12_INDEX,
	APU_PARAMETER_RATE_STEP_Q16_INDEX,
	APU_RATE_STEP_Q16_ONE,
	type ApuParameterRegisterWords,
} from './contracts';

export type ApuFilterType = BiquadFilterType;

export interface ApuOutputFilter {
	type: ApuFilterType;
	frequency: number;
	q: number;
	gain: number;
}

export interface ApuOutputPlayback {
	playbackRate: number;
	gainLinear: number;
	filter: ApuOutputFilter | null;
}

export function resolveApuGainLinear(gainQ12Word: number): number {
	return toSignedWord(gainQ12Word) / APU_GAIN_Q12_ONE;
}

export function resolveApuPlaybackRate(rateStepQ16Word: number): number {
	return toSignedWord(rateStepQ16Word) / APU_RATE_STEP_Q16_ONE;
}

function decodeApuFilterType(kind: number): ApuFilterType {
	switch (kind) {
		case APU_FILTER_HIGHPASS:
			return 'highpass';
		case APU_FILTER_BANDPASS:
			return 'bandpass';
		case APU_FILTER_NOTCH:
			return 'notch';
		case APU_FILTER_ALLPASS:
			return 'allpass';
		case APU_FILTER_PEAKING:
			return 'peaking';
		case APU_FILTER_LOWSHELF:
			return 'lowshelf';
		case APU_FILTER_HIGHSHELF:
			return 'highshelf';
		default:
			return 'lowpass';
	}
}

export function resolveApuOutputFilter(registerWords: ApuParameterRegisterWords): ApuOutputFilter | null {
	const filterKind = registerWords[APU_PARAMETER_FILTER_KIND_INDEX]!;
	if (filterKind === APU_FILTER_NONE) {
		return null;
	}
	return {
		type: decodeApuFilterType(filterKind),
		frequency: toSignedWord(registerWords[APU_PARAMETER_FILTER_FREQ_HZ_INDEX]!),
		q: toSignedWord(registerWords[APU_PARAMETER_FILTER_Q_MILLI_INDEX]!) / 1000,
		gain: toSignedWord(registerWords[APU_PARAMETER_FILTER_GAIN_MILLIDB_INDEX]!) / 1000,
	};
}

export function resolveApuOutputPlayback(registerWords: ApuParameterRegisterWords): ApuOutputPlayback {
	return {
		playbackRate: resolveApuPlaybackRate(registerWords[APU_PARAMETER_RATE_STEP_Q16_INDEX]!),
		gainLinear: resolveApuGainLinear(registerWords[APU_PARAMETER_GAIN_Q12_INDEX]!),
		filter: resolveApuOutputFilter(registerWords),
	};
}

File: src/bmsx_cpp/machine/devices/audio/playback.cpp
#include "machine/devices/audio/playback.h"

#include "machine/common/numeric.h"

namespace bmsx {

f32 resolveApuGainLinear(u32 gainQ12Word) {
	return static_cast<f32>(toSignedWord(gainQ12Word)) / static_cast<f32>(APU_GAIN_Q12_ONE);
}

f32 resolveApuPlaybackRate(u32 rateStepQ16Word) {
	return static_cast<f32>(toSignedWord(rateStepQ16Word)) / static_cast<f32>(APU_RATE_STEP_Q16_ONE);
}

std::optional<ApuOutputFilter> resolveApuOutputFilter(const ApuParameterRegisterWords& registerWords) {
	const u32 filterKind = registerWords[APU_PARAMETER_FILTER_KIND_INDEX];
	if (filterKind == APU_FILTER_NONE) {
		return std::nullopt;
	}
	ApuOutputFilter filter;
	switch (filterKind) {
		case APU_FILTER_HIGHPASS:
			filter.type = "highpass";
			break;
		case APU_FILTER_BANDPASS:
			filter.type = "bandpass";
			break;
		case APU_FILTER_NOTCH:
			filter.type = "notch";
			break;
		case APU_FILTER_ALLPASS:
			filter.type = "allpass";
			break;
		case APU_FILTER_PEAKING:
			filter.type = "peaking";
			break;
		case APU_FILTER_LOWSHELF:
			filter.type = "lowshelf";
			break;
		case APU_FILTER_HIGHSHELF:
			filter.type = "highshelf";
			break;
		default:
			filter.type = "lowpass";
			break;
	}
	filter.frequency = static_cast<f32>(toSignedWord(registerWords[APU_PARAMETER_FILTER_FREQ_HZ_INDEX]));
	filter.q = static_cast<f32>(toSignedWord(registerWords[APU_PARAMETER_FILTER_Q_MILLI_INDEX])) / 1000.0f;
	filter.gain = static_cast<f32>(toSignedWord(registerWords[APU_PARAMETER_FILTER_GAIN_MILLIDB_INDEX])) / 1000.0f;
	return filter;
}

ApuOutputPlayback resolveApuOutputPlayback(const ApuParameterRegisterWords& registerWords) {
	ApuOutputPlayback playback;
	playback.playbackRate = resolveApuPlaybackRate(registerWords[APU_PARAMETER_RATE_STEP_Q16_INDEX]);
	playback.gainLinear = resolveApuGainLinear(registerWords[APU_PARAMETER_GAIN_Q12_INDEX]);
	playback.filter = resolveApuOutputFilter(registerWords);
	return playback;
}

} // namespace bmsx
